File: letscode/diffsort/mergesort/merge.py
# coding=utf-8
def sort(input):
    if not input or not isinstance(input, list) or len(input) <= 0:
        return input
    output = merge(input)
    return output

def merge(input):
    if len(input) <= 1:
        return input
    output = []
    mid = len(input) / 2
    left = merge(input[:mid])
    right = merge(input[mid : len(input)])
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            output.append(left[i])
            i += 1
        else:
            output.append(right[j])
            j += 1
    while i < len(left):
        output.append(left[i])
        i += 1
    while j < len(right):
        output.append(right[j])
        j += 1
    return output

if __name__ == '__main__':
    input = [6, 5, 7, 8, 1, 3, 2, 4, 0]
    print (sort(input))

File: letscode/diffsort/mergesort/merge.go
package main

import "fmt"

func merge(input []int) []int {
	if len(input) <= 1 {
		return input
	}
	output := make([]int, 0, len(input))
	n := len(input) / 2
	left := merge(input[:n])
	right := merge(input[n:])
	i := 0
	j := 0
	for i < len(left) && j < len(right) {
		if left[i] < right[j] {
			output = append(output, left[i])
			i++
		} else {
			output = append(output, right[j])
			j++
		}
	}
	for i < len(left) {
		output = append(output, left[i])
		i++
	}
	for j < len(right) {
		output = append(output, right[j])
		j++
	}
	return output
}

func main() {
	input := []int{6, 5, 7, 8, 1, 3, 2, 4, 0}
	output := merge(input)
	fmt.Println(output)
}
